### ISSUE TITLE
Simplify API

### ISSUE THREAD

**Issue by mottosso** [pyblish/pyblish-base#170]
# Goal

To simplify the API. :) 
# Implementation

More precisely, to move some of the logic from `pyblish.plugin` to `pyblish.util` where it belongs. Mainly, `pyblish.plugin.Plugin.process()` which takes a context and goes to work publishing. It yields instances and optional errors.

The behaviour is good, but it's place in the API is not. Currently, the Pyblish QML GUI jumps over it straight into the `process_*` functions which goes to show that `process()` isn't really needed by anything but the headless `pyblish.util.publish()` helper function.
### Coroutines

There should still be a method of unifying which plug-ins and instances are selected for processing, possibly as a coroutine.

The behaviour involved in processing the appropriate pair of (plugin, instance) is non-trivial and critical to bridging the behaviour between the various frontends. We could encapsulate it in a coroutine, but the problem is how to transfer this knowledge over to a client via TCP, that may not even understand Python.

One option might be to transmit items of interest, and pre-process their order in the backend only to  send back a list of (plugin, instance) pairs. With this option, it would be possible to use the same coroutine for any frontend, at the expense of an extra TCP call and a deserialisation/serialisation process.
# Future

Implementing this makes for more expressive frontends, and leaves room for the future Dependency Injection proposal (#127).

**Comment by mottosso** [comment]
To get explicit, it's this type of logic that doesn't belong in the API.

``` python
if instance.has_data("publish"):
    if instance.data("publish", default=True) is False:
        self.log.debug("Skipping %s, "
                       "publish-flag was false" % instance)
        continue
```
- [Line 212](https://github.com/pyblish/pyblish/blob/c1be153f9fbeb3f9456e3213b8ce7902beae91a2/pyblish/plugin.py#L212)

**Comment by mottosso** [comment]
Implemented in 1.1
